### PR TITLE
Add `fix_external` to randomization functions

### PR DESCRIPTION
**Pull request by siyuzhou**
A quick and dirty modification to the original implementation. New proposed API still in work.

### CODE REVIEW

File: neet/boolean/randomnet.py [hunk 1]
```diff
@@ -141,64 +158,69 @@ def _random_logic_free_connections(logic_net, ps):
...
 
-        if n_indices > 0:
-            for state in StateSpace(n_indices):
-                if random.random() < ps[i]:
-                    conditions.add(tuple(state))
+    internals = [idx for idx in range(logic_net.size) if idx not in externals]
+    num_internal_connections = np.zeros(len(internals))
 
-        new_table.append((indices, conditions))
+    sample = np.random.choice([i // logic_net.size for i in range(len(internals) * logic_net.size)],
```
Comment (bcdaniels) [inline neet/boolean/randomnet.py]: Clever!  Better than the way I was doing it before.

File: neet/boolean/randomnet.py [hunk 2]
```diff
@@ -66,7 +66,15 @@ def _random_binary_states(n, p):
     return set(tuple(state) for state in StateSpace(n) if random.random() < p)
 
 
-def _random_logic_fixed_connections(logic_net, ps):
+def _external_nodes(logic_net):
+    externals = set()
+    for idx, row in enumerate(logic_net.table):
+        if len(row[0]) == 1 and row[0][0] == idx:
```
Comment (dglmoore) [inline neet/boolean/randomnet.py]: Is this equivalent to `if row[0] == [idx]:`?

File: neet/boolean/randomnet.py [hunk 3]
```diff
@@ -66,7 +66,15 @@ def _random_binary_states(n, p):
     return set(tuple(state) for state in StateSpace(n) if random.random() < p)
 
 
-def _random_logic_fixed_connections(logic_net, ps):
+def _external_nodes(logic_net):
```
Comment (dglmoore) [inline neet/boolean/randomnet.py]: This function defines a node as external if and only if it has exactly one incoming edge, that edge is a self-edge, and that self-edge represents the "identify" function (0 -> 0, 1 -> 1). Is this correct? My concern is whether or not we can have networks with nodes that meet this criterion, but aren't listed in the "external.txt" file as an external edge.

File: neet/boolean/randomnet.py (hunk 2 above)
Comment (siyuzhou) [reply neet/boolean/randomnet.py]: That's right. I'll change that.

File: neet/boolean/randomnet.py (hunk 3 above)
Comment (siyuzhou) [reply neet/boolean/randomnet.py]: I thought about this possibility when I wrote this. But the "external" nodes are stored in the LogicNetwork just as any other nodes. We have to give `LogicNetwork` an attribution `externals` and store external nodes explicitly to preserve the difference. Then we have to consider adding the same attribute the `WTNetwork` as there is no reason why `LogicNetwork` gets this special treatment. So before we do that, I convinced myself that when  they built the biological networks, any node that would have logic equation A = A is already put in the "external.txt" file. But I admit this assumption is likely wrong.

File: neet/boolean/randomnet.py (hunk 2 above)
Comment (siyuzhou) [reply neet/boolean/randomnet.py]: And I need to check whether the condition is `{'1'}`.

File: neet/boolean/randomnet.py (hunk 3 above)
Comment (siyuzhou) [reply neet/boolean/randomnet.py]: I did a quick check. There are indeed some networks which have nodes with expression A = A but not included in the "expression.txt" file.

**Comment by dglmoore** [comment]
Perhaps we should extend the various network classes to include a member
variable which specifies which nodes are external?

**Pull request merged**